hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/utils/AzcopyToolHelper.java

```diff
@@ -0,0 +1,274 @@
+import static org.apache.hadoop.fs.azurebfs.constants.AbfsHttpConstants.FORWARD_SLASH;
+import static org.apache.hadoop.fs.azurebfs.constants.FileSystemUriSchemes.ABFS_BLOB_DOMAIN_NAME;
+import static org.apache.hadoop.fs.azurebfs.constants.FileSystemUriSchemes.ABFS_DFS_DOMAIN_NAME;
+
+/**
+ * Singleton class to create a file or folder in Azure Blob Storage using Azcopy tool.
+ * <a href="https://learn.microsoft.com/en-us/azure/storage/common/storage-use-azcopy-v10">
+ * Azcopy</a> is a command-line utility tool to copy blobs or files to or from a storage account.
+ * It uses Blob Endpoint and ends up creating implicit paths in the storage account.
+ * We will leverage this tool to create implicit paths in storage account for testing purposes.
+ */
+public class AzcopyToolHelper {
+
+  private File hadoopAzureDir;
+  private String azcopyDirPath;
+  private String azcopyExecutablePath;
+  private String fileCreationScriptPath;
+  private String folderCreationScriptPath;
+  private String sasToken;
+
+  private static final String USER_DIR_SYSTEM_PROPERTY = "user.dir";
+  private static final String HADOOP_AZURE_DIR = "hadoop-azure";
+  private static final String AZCOPY_DIR_NAME = "azcopy";
+  private static final String AZCOPY_EXECUTABLE_NAME = "azcopy";
+  private static final String FILE_CREATION_SCRIPT_NAME = "createAzcopyFile";
+  private static final String FOLDER_CREATION_SCRIPT_NAME = "createAzcopyFolder";
+  private static final String DIR_NOT_FOUND_ERROR = " directory not found";
+  private static final String AZCOPY_DOWNLOADED_DIR_NAME = "/azcopy_linux_amd64_*/* ";
+  private static final String AZCOPY_DOWNLOADED_TAR_NAME = "/azcopy_linux_amd64_* azcopy.tar.gz";
+
+  private static final String AZCOPY_CMD_SHELL = "bash";
+  private static final String AZCOPY_CMD_OPTION = "-c";
+  private static final String AZCOPY_DOWNLOAD_URL = "https://aka.ms/downloadazcopy-v10-linux";
+  private static final String AZCOPY_DOWNLOAD_CMD = "wget " + AZCOPY_DOWNLOAD_URL + " -O azcopy.tar.gz" + " --no-check-certificate";
+  private static final String EXTRACT_CMD = "tar -xf azcopy.tar.gz -C ";
+  private static final String MOVE_CMD = "mv ";
+  private static final String REMOVE_CMD = "rm -rf ";
+  private static final String CHMOD_CMD = "chmod +x ";
+  private static final char QUESTION_MARK = '?';
+  private static final ReentrantLock LOCK = new ReentrantLock();
+
+  // Same azcopy tool can be used for all the tests running in save JVM.
+  private static AzcopyToolHelper azcopyToolHelper = null;
+
+  private AzcopyToolHelper() {
+
+  }
+
+  /**
+   * Constructor to initialize the AzcopyToolHelper.
+   * Azcopy tool work with SAS based authentication. SAS can be configured using
+   * test configuration "fs.azure.test.fixed.sas.token".
+   * @param sasToken to be used for authentication.
+   */
+  public static AzcopyToolHelper getInstance(String sasToken) throws IOException, InterruptedException {
+    if (azcopyToolHelper == null) {
```

Comment:
This double checked locking mechanism can be replaced with Bill Pugh Singleton Design. For example :-

private static class HelperHolder {
    private static final AzcopyToolHelper INSTANCE = new AzcopyToolHelper();
  }

  public static AzcopyToolHelper getInstance(String sasToken) throws IOException, InterruptedException {
    AzcopyToolHelper instance = HelperHolder.INSTANCE;
    instance.init(sasToken);
    return instance;
  }

Reply:
Nice Suggestion.
Taken

Reply:
We actually do not need a forwarding class. Reason being, the constructor of AzcopyHelper is not compute-heavy / IO-heavy and also, since its empty, there is no exception that would be raised from the constructor.

As far as singleton initialization is concerned, that would be taken care by the ClassLoader. Even in the suggested way, multiple threads can enter `AzcopyToolHelper instance = HelperHolder.INSTANCE;` which would want to intialize HelperHolder which would initialize AzcopyHelper. Now even though multiple threads want to create HelperHolder, we would still depend on classLoader to create a singleton for static methods/fields. So, if we are dependent on ClassLoader, lets not have an additional class :).

Hence, I would kindly suggest that we keep the code we have right now. Thanks!